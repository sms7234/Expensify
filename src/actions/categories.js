import database from '../firebase/firebase';

export const setCategories = (categories) => ({
  type:"SET_CATEGORIES",
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`)
      .once('value')
      .then((snapshot) => {
        const categories = [];
        snapshot.forEach((child) => {
          categories.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setCategories(categories));
      });
  };
};

export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) =>{
    const {
      category= '',
      description= ''
    } = categoryData;
    const categoryItem = {category, description};
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`).push(categoryItem).then((ref)=>{
      dispatch(addCategory({
        id: ref.key,
        ...categoryItem
      }));
    });
  };
};

export const editCategory = (id, updates) => ({
  type: 'EDIT_CATEGORY',
  id,
  updates
});

export const startEditCategory = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).update(updates).then(()=>{
      dispatch(editCategory(id,updates));
    });
  };
};

export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
});

export const startRemoveCategory = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).remove().then(()=>{
      dispatch(removeCategory(id));
    });
  };
};
