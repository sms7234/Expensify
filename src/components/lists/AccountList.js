import React from  'react';
import {connect} from 'react-redux';
import TwoDataFieldListItem from '../items/TwoDataFieldListItem';
import selectAccounts from '../../selectors/accounts';


export const AccountList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Accounts</div>
      <div className="show-for-desktop">Accounts</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.accounts.length === 0 ? (
        <div className="list-item--message">
          <span>No accounts to show</span>
        </div>
      ): (
        props.accounts.map((item) => {return <TwoDataFieldListItem
       key={item.id}
       id={item.id}
       value={item.account}
       description={item.description}
       location={'editAccount'}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    accounts: selectAccounts(state.accounts, state.filters)
  };
};

export default connect(mapStateToProps)(AccountList);
