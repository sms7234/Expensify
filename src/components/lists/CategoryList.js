import React from  'react';
import {connect} from 'react-redux';
import CategoryListItem from '../items/CategoryListItem';
import selectCategories from '../../selectors/categories';


export const CategoryList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Categories</div>
      <div className="show-for-desktop">Categories</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.categories.length === 0 ? (
        <div className="list-item--message">
          <span>No categories to show</span>
        </div>
      ): (
        props.categories.map((item) => {return <CategoryListItem key={item.id}{...item}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    categories: selectCategories(state.categories, state.filters)
  };
};

export default connect(mapStateToProps)(CategoryList);
