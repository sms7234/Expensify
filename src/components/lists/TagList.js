import React from  'react';
import {connect} from 'react-redux';
import TwoDataFieldListItem from '../items/TwoDataFieldListItem';
import selectTags from '../../selectors/tags';


export const TagList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Tags</div>
      <div className="show-for-desktop">Tags</div>
      <div className="show-for-desktop">Description</div>
    </div>
    <div className="list-body">
    {
      props.tags.length === 0 ? (
        <div className="list-item--message">
          <span>No tags to show</span>
        </div>
      ): (
        props.tags.map((item) => {return <TwoDataFieldListItem
       key={item.id}
       id={item.id}
       value={item.tag}
       description={item.description}
       location={'editTag'}/>;})
      )
    }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    tags: selectTags(state.tags, state.filters)
  };
};

export default connect(mapStateToProps)(TagList);
