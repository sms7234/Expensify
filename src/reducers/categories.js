export default (state = [], action) => {
  switch(action.type){
    case 'ADD_CATEGORY':
      return state.concat(action.category);
    case 'SET_CATEGORIES':
      return action.categories;
    default:
      return state;
  }
};
