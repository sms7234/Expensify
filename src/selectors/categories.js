export default (categories, {category}) => {
  return categories.filter((item) => {
    const textMatch = item.category.toLowerCase().includes(category.toLowerCase());
    return textMatch;
  }).sort((a,b)=>{
    if(a.category < b.category){
      return -1;
    } else {
      return 1;
    }
  });
};
