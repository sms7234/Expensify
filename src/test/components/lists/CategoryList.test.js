import React from 'react';
import {shallow} from 'enzyme';
import {CategoryList} from '../../../components/lists/CategoryList';
import categories from '../../fixtures/categories';

test('should render CategoryList with categories', () => {
  const wrapper = shallow(<CategoryList categories={categories} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render CategoryList with no categories', () => {
  const wrapper = shallow(<CategoryList categories={[]} />);
  expect(wrapper).toMatchSnapshot();
})
