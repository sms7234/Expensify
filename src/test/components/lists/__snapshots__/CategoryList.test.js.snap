// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CategoryList with categories 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Categories
    </div>
    <div
      className="show-for-desktop"
    >
      Categories
    </div>
    <div
      className="show-for-desktop"
    >
      Description
    </div>
  </div>
  <div
    className="list-body"
  >
    <CategoryListItem
      description="dog costs"
      id="1"
      key="1"
      location="editCategory"
      value="Dog"
    />
    <CategoryListItem
      description="food costs"
      id="2"
      key="2"
      location="editCategory"
      value="Groceries"
    />
    <CategoryListItem
      description="car costs"
      id="3"
      key="3"
      location="editCategory"
      value="Car"
    />
  </div>
</div>
`;

exports[`should render CategoryList with no categories 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Categories
    </div>
    <div
      className="show-for-desktop"
    >
      Categories
    </div>
    <div
      className="show-for-desktop"
    >
      Description
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item--message"
    >
      <span>
        No categories to show
      </span>
    </div>
  </div>
</div>
`;
